### PR TITLE
New pluggable metrics engine

### PR DESCRIPTION
**Pull request by JoseDanielGar**
El folder metrics implementa un sistema de métricas enchufables. En core/ están las clases base (Metric, PromptRow, AudioItem) y el registro central. En implementation/ vive la métrica CLAP que calcula similitud texto–audio y genera un CSV DVC-friendly y un resumen JSON. Las métricas FAD y Aesthetics quedan como stubs no implementados (fail-fast) para habilitar su futura activación sin cambiar llamadas.

El folder evaluation contiene un motor simple que lee config.yaml, carga los prompts desde data/prompts/spanio_prompts.csv y empareja audios por id en data/tracks/generated_music. Con el registro de métricas ejecuta solo las habilitadas en la configuración, escribiendo resultados en data/scores/ y un combined_summary.json. Respeta environment.device tal cual (p. ej., "auto").

Para correr la evaluación: uv run -m models.scripts.evaluate. Actualmente produce data/scores/results_with_clap.csv, clap_summary.json y combined_summary.json. Para añadir otra métrica, basta crear su implementación, registrarla y declararla en evaluation.metrics del config.yaml.

### CODE REVIEW

**Review by jlsarmientoh — APPROVED** [review]

File: config.yaml [hunk 1]
```diff
@@ -56,4 +56,10 @@ logging:
...
   num_workers: 4
-  pin_memory: true
\ No newline at end of file
+  pin_memory: true
+
+# Evaluation configuration
+evaluation:
+  metrics:
+    clap:
+      output_csv_name: "results_with_clap.csv"
```
Comment (mreyesoatuniandes) [inline config.yaml]: Añadir blank line al final del archivo.

File: config/__init__.py
Comment (mreyesoatuniandes) [inline config/__init__.py]: Este package y su __init__.py tienen referencias extrañas (relativas) en importaciones.
Podemos revisarlo luego.

File: config/types.py [hunk 2]
```diff
@@ -1,5 +1,6 @@
 from dataclasses import dataclass, field
-from typing import Any, Dict
```
Comment (mreyesoatuniandes) [inline config/types.py]: Puedes cambiar la referencia de `Dict` -> dict en todo este archivo?
Estamos usando Python 3.11, no es necesario usar el support de `Dict` o `List`

File: config/manager.py [hunk 3]
```diff
@@ -67,14 +70,20 @@ def _create_config_objects(self, config_data: Dict[str, Any]) -> Config:
             inference_config = InferenceConfig(**config_data.get('inference', {}))
             logging_config = LoggingConfig(**config_data.get('logging', {}))
             environment_config = EnvironmentConfig(**config_data.get('environment', {}))
-            
+            # Evaluation dataclasses
```
Comment (mreyesoatuniandes) [inline config/manager.py]: Cual es el caso para usar `... or {}`? 
Cuando se hace get para el caso cuando no esta la llave se inicializa un dict.
Podria ser no necesario y simple solo usando `.get()`

Puedes ajustarlo en las ocurrencias que tiene este archivo?

File: metrics/core/base.py
Comment (mreyesoatuniandes) [inline metrics/core/base.py]: Tal ves sea mejor definir esto como un `typedefs.py`

File: metrics/clap.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        # Load exactly like the legacy script
+        wav, sr = torchaudio.load(it.audio_path)  # [C, T] as-is
+        if sr != TARGET_SR:
+            wav = torchaudio.functional.resample(wav, sr, TARGET_SR)
+        wav = wav.to(dev)
+
+        aemb = model.get_audio_embedding_from_data(wav, use_tensor=True)
+        temb = model.get_text_embedding([it.description], use_tensor=True)
+        aemb = torch.nn.functional.normalize(aemb, dim=-1)
+        temb = torch.nn.functional.normalize(temb, dim=-1)
```
Comment (mreyesoatuniandes) [inline metrics/clap.py]: Por qué se aplica normalización?

File: metrics/clap.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+def _resolve_device(device: Optional[str]) -> str:
+    # Match legacy behavior: choose CUDA if available, else CPU.
+    # We intentionally do not propagate "auto" to torch.
+    if device in (None, "", "auto"):
+        return "cuda" if torch.cuda.is_available() else "cpu"
+    return device
+
+
+def _load_clap(device: str) -> CLAP_Module:
+    model = CLAP_Module(enable_fusion=True)
```
[the remark this answers is not in the document]
Comment (JoseDanielGar) [reply metrics/clap.py]: https://github.com/audioreprompt-org/audio_reprompt/issues/13

File: metrics/clap.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+    out: List[CLAPScored] = []
+    for it in items:
+        # Load exactly like the legacy script
+        wav, sr = torchaudio.load(it.audio_path)  # [C, T] as-is
+        if sr != TARGET_SR:
+            wav = torchaudio.functional.resample(wav, sr, TARGET_SR)
+        wav = wav.to(dev)
+
+        aemb = model.get_audio_embedding_from_data(wav, use_tensor=True)
```
[the remark this answers is not in the document]
Comment (JoseDanielGar) [reply metrics/clap.py]: No entendí bien el comentarío, podrías profundizar?

[the remark this answers is not in the document]
Comment (mreyesoatuniandes) [reply metrics/clap.py]: La documentacion (link) indica otra forma para obtener los embeddings de una pieza de audio.
Mi sugerencia inicial es revisar si el proceso actual vs el proceso que tiene la documentación genera los mismos embeddings.

**Review by mreyesoatuniandes — CHANGES_REQUESTED** [review]
@JoseDanielGar observé algunos de los cambios y otros comentarios no fueron respondidos o direccionados. Al final dejo los cambios de estructura a tu decisión.

La acción principal que necesitamos en este PR es asegurar la implementación del CLAPScore, sugiero en los comentarios revisar una implementación analoga y comparar. 

Debido a que esta métrica es bastante critica para las tareas dependientes, sugiero escribir un par de unit tests con el fin de cubrir el resultado esperado. 
Pueden ser dos o tres que cubran casos de generación del CLAP y comparación entre audio y text embeddings, sin dependencia de datos de DVC, puedes utilizar datos sinteticos con torchaudio.

File: metrics/clap.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+@dataclass
+class CLAPItem:
+    id: str
+    description: str
+    audio_path: str
+    instrument: Optional[str] = None
+
+
+@dataclass
+class CLAPScored(CLAPItem):
```
[the remark this answers is not in the document]
Comment (JoseDanielGar) [reply metrics/clap.py]: Disculpa, para este pense que había respondido. A partir del comentario, cambié la relación entre CLAPItem y CLAPScored de herencia a composición. Ahora CLAPScored contiene un CLAPItem en lugar de extenderlo, dejando clara la separación entre los datos de entrada y los resultados del cálculo. Además, marqué ambas dataclasses como frozen=True para garantizar inmutabilidad

File: metrics/clap.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (JoseDanielGar) [reply metrics/clap.py]: Dejé los tres backends por si quieres probar, y un mini script:

 uv run -m pytest metrics/tests/test_clap_backends.py -q

File: metrics/clap/backends/hf.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+            from transformers import ClapProcessor, ClapModel  # type: ignore
+        except Exception as e:
+            raise BackendUnavailable("Install `transformers` to use hf_processor backend.") from e
+        model_id = (backend_cfg or {}).get("model_id", "laion/clap-htsat-fused")
+        try:
+            self.processor = ClapProcessor.from_pretrained(model_id)
+            self.model = ClapModel.from_pretrained(model_id)
+            self.model.eval().to(self.device)
+        except Exception as e:
+            raise BackendUnavailable(f"Failed to load HF model '{model_id}'.") from e
```
Comment (mreyesoatuniandes) [inline metrics/clap/backends/hf.py]: La idea de tener backend es interesante.
Debemos asegurar que todos usen la misma fuente de pesos, que habia sido indicada antes.
[music_audioset_epoch_15_esc_90.14.pt](https://huggingface.co/lukewys/laion_clap/blob/main/music_audioset_epoch_15_esc_90.14.pt)

File: metrics/clap/backends/laion.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    name = "laion_module"
+
+    def __init__(self, device: str, backend_cfg: Optional[dict[str, Any]] = None) -> None:
+        self.device = _resolve_device(device)
+        try:
+            from laion_clap import CLAP_Module  # type: ignore
+        except Exception as e:
+            raise BackendUnavailable("Install `laion-clap` to use laion_module backend.") from e
+        self.model = CLAP_Module(enable_fusion=True)
+        self.model.load_ckpt()
```
Comment (mreyesoatuniandes) [inline metrics/clap/backends/laion.py]: Mismo comentario anterior.

File: metrics/clap/types.py
Comment (mreyesoatuniandes) [inline metrics/clap/types.py]: cambiar a `typedefs.py`

File: metrics/core/base.py [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+    def run(
+            self,
+            prompts: list[PromptRow],
+            audio_items: list[AudioItem],
+            metric_cfg: dict[str, Any],
+            device: str,
+            scores_dir: Path
+    ) -> dict[str, Any]:
+        raise NotImplementedError
```
Comment (mreyesoatuniandes) [inline metrics/core/base.py]: Esta no podría ser definida como una ABC?

File: metrics/tests/test_clap_backends.py [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+                continue
+            wav_path = tracks_dir / f"{rid}.wav"
+            if not wav_path.exists():
+                continue
+            desc = str(r.get("description") or "").strip()
+            instr = str(r.get("instrument") or "").strip() or None
+            items.append(CLAPItem(id=rid, description=desc, audio_path=str(wav_path), instrument=instr))
+            if len(items) >= max_items:
+                break
+    return items
```
Comment (mreyesoatuniandes) [inline metrics/tests/test_clap_backends.py]: Por qué estamos usando dependencias sobre datos reales en estos unit-tests?
Si quiseramos colocar un Github Actions (GA) esto implica dependencias. Me parece util tenerlo como un script externo para usar como validación como spot checks, pero no como unit-tests.

Puedes dirigir los cambios?

**Review by mreyesoatuniandes — APPROVED** [review]
Deje comentarios para que puedas aplicar cambios.
Recomiendo adjuntar una prueba de la comparacion, tipo un CSV con los backends implementados en la conversación del PR antes de mezclar los cambios.
